### ISSUE TITLE
Redistribution request for NixOS

### ISSUE THREAD

**Issue by Baughn** [bolthole/zrep#14]
Hi,

I'd like to package zrep for NixOS. (https://nixos.org)

Because of the nature of this distribution, I'll need to make modifications. This will be things like changing the interpreter path to /usr/bin/env ksh, and otherwise adapting it to NixOS's nonstandard directory structure. Additionally we'd like to autogenerate a "compiled" version for the Nix derivation cache, so there'd be redistribution as well.

Is this okay with you?

On a secondary note, which exact version of ksh do you need to use? The link on http://www.kornshell.com/software/ is dead.

Oh, and while you're looking at this... is it okay to use zrep on a parent filesystem, where I don't want to replicate its children?

**Comment by ppbrown** [comment]
Hello,

this sounds in principle like something I am in favour of.
There are some details I would like to work out with you first.

#1 would be the portability question. I would like to know the full scope
of modifications needed to make zrep compatible with nixOS, so that I can
potentially merge them into the main zrep source.
So, I would like more details about this "nonstandard directory structure".
I'd also like to know what you mean by a compiled version.

As far as what version of ksh: ksh93 or later, as far as I am aware.
It's too bad that korgshell.com link is decayed. Seems like the most
reliable place to get actual source code now, is
https://github.com/att/ast
(since korn shell is considered part of the AT&T "ast-open" software
distribution)

**Comment by Baughn** [comment]
The only change I can state up front as 'needs to happen' is the use of #!/usr/bin/env ksh instead of #!/bin/ksh. NixOS doesn't currently have ksh packaged, and I didn't want to spend time doing so unless I had in-principle agreement, but I can talk you through the installation script for zrep before merging and you can decide what, if any changes, to make upstream.

Slight digression. I mentioned that NixOS does not follow the filesystem hierarchy standard; in fact /lib and /sbin do not exist, while the only standard binaries that exist are /bin/sh and /usr/bin/env. /usr is otherwise empty. The reason for this is to prevent packages from using dependencies which they haven't declared.

Instead, all packages are stored in per-package directories in /nix/store. This is a recursive process; packages with dependencies reference the paths of those dependencies, directly, which leads to packages forming a tree-structure with the "current system" at the tip of the tree. That package is actually referenced from /run/current-system, which is where PATH and other envvars point to, so upgrades are atomic. The system is garbage-collected; by default, if something breaks you can rollback. If it breaks *badly*, you can reboot and ask for an earlier generation.

That's probably more information than you really want, though. I don't want to take up too much of your time, so... relevant facts:
- Although the zfs binaries exist on the user's path, installed shell scripts (and anything else) are by-default wrapped such that they cannot see anything they haven't explicitly asked for. This is a feature; it means, for example, that multiple packages whose dependencies clash do not necessarily interfere with each other.
- Since the FHS paths are empty, any absolute paths hardcoded into the script won't work.
- The default package installer runs a 'fixup' script, which will rewrite such absolute paths to point at the actual location of the binaries. The paths aren't fixed; they include the hash of every input to that package, including its dependencies, so a tiny change somewhere in libc will make the path to zfs (or zpool) change. Normally you don't need to care.

And:
- There exists a cache server, which attempts to build all packages with default options. It doesn't matter much for something as light as a shell script, but in principle everything is compiled from source; as Nix intends to have repeatable builds, downloading the precompiled packages is strictly an optimization. For zrep such a download would include all the fixups already applied.

**Comment by ppbrown** [comment]
What about "absolute paths" to things like /var/run and /var/tmp ?

**Comment by Baughn** [comment]
/var does exist, and is used in the normal fashion.

**Comment by Baughn** [comment]
Oh yeah. Also, could you version the script download?

Since Nix doesn't currently mirror its sources, but does include a hash in the package definition, it means that any time you change zrep the package will break.

**Comment by ppbrown** [comment]
oh okay.
i already had a secret versioning :)

for now. just use  www.bolthole.com/solaris/zrep/zrep.1.5.4

in future I'll make the link public on the page

**Comment by Baughn** [comment]
It seems like I'll have to shelve this project, for now. I can't get ksh to build, and it's not like there's an active developer community to ask.

Just for the record, I suppose:
'''
/tmp/nix-build-ksh-93.20120801.drv-0/ast-3f54fd6/src/cmd/kshlib/dbm/dbm.c:32:2: error: unknown type name 'DBM'
'''

**Comment by ppbrown** [comment]
You could always try the redhat source package instead
http://ftp.redhat.com/pub/redhat/linux/enterprise/6Server/en/os/SRPMS/

**Comment by Baughn** [comment]
It gives the same error. There's some kind of problem with my environment, but since there's no documentation of what it needs, and no obvious error messages, I'm unlikely to figure out what that might be.

**Comment by ppbrown** [comment]
Maybe yhou should try an older version. like the 2001. instead of the 2012
source. as mentioned in the ksh book, compiling section:

http://docstore.mik.ua/orelly/unix3/korn/appc_03.htm

according to that page it shoud compile out of the box, no tweaks, on linux.

that src version is available from slackware.

http://ftp.slackware.com/pub/slackware/slackware-9.0/source/ap/ksh93/

**Comment by Baughn** [comment]
Took a look at that. It seems even harder to build, as it hardcodes the paths of all the basic tools; /bin/cp, etc. The newer version got a lot further, even if it failed in the end.

**Comment by Baughn** [comment]
I don't suppose your script will work with bash?

**Comment by ppbrown** [comment]
it will not, without a fair chunk of rewriting stuff  like [[ ]]

**Comment by Baughn** [comment]
Yes, all right. I also ran into some other problems with replication. i.e. unstable network connections, soo~ I think I'll end up writing my own. Yours looks great, but I could probably have gotten something working already if I wasn't spending time on ksh. :X

**Comment by ppbrown** [comment]
Hmm. well,lt's your choice... but personally, I think you would best spend
your time fixing the issue in NixOS, where
stuff that compiles for other linuxen, doesnt compile on NixOS!
:-}

**Comment by Baughn** [comment]
We could discuss that for days... but let's not do it here. If you get curious enough to actually want to know my reply to that one, try IRC. :P

**Comment by ppbrown** [comment]
well, i'm on the nixos channel now.. where are you ? !

**Comment by ppbrown** [comment]
A more direct update.
I dont know how you tried to build ksh,but you did it wrong :-}
I followed the directions in github and it worked easly.

From memory, this is what I did:

(install nixos VM)
nix-env -i gnumake  gcc-wrapper-4.9.3 bison yacc git 
git clone https://github.com/att/ast.git
cd ast
sh ./bin/package/make

and then it ground away building the tree. It's only halfway through at the moment, but it successfully built  arch/linux.i386/bin/ksh
and I can start it
